common/txmgr/tracker.go

```diff
@@ -0,0 +1,247 @@
+	tr.txCache = map[int64]AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	tr.enabledAddrs = map[ADDR]bool{}
+}
+
+// HandleAbandonedTxes is called by the Confirmer to update abandoned transactions
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) HandleAbandonedTxes(ctx context.Context) error {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	if !tr.isTracking {
+		return errors.New("not tracking abandoned txes")
+	}
+
+	for id, atx := range tr.txCache {
+		if finalized := tr.finalizeTx(ctx, atx); finalized {
+			delete(tr.txCache, id)
+		}
+	}
+	return nil
+}
+
+// GetAbandonedAddresses returns list of abandoned addresses being tracked
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) GetAbandonedAddresses() []ADDR {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	if !tr.isTracking {
+		tr.lggr.Warn("not tracking abandoned txes")
+	}
+
+	addrs := make([]ADDR, len(tr.txCache))
+	for _, atx := range tr.txCache {
+		if atx.isValid() {
+			addrs = append(addrs, atx.fromAddress)
+		}
+	}
+	return addrs
+}
+
+// insertTx inserts a transaction into the tracker as an AbandonedTx
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) insertTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	if _, contains := tr.txCache[tx.ID]; contains {
+		return
+	}
+
+	tr.txCache[tx.ID] = AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{
+		id:          tx.ID,
+		fromAddress: tx.FromAddress,
+		fatalTime:   time.Now().Add(tr.ttl),
+	}
+	tr.lggr.Debugw(fmt.Sprintf("inserted tx %v", tx.ID))
+}
+
+// finalizeTx tries to finalize a transaction based on its current state.
+// Transactions exceeding ttl are marked fatal.
+// Returns true if the transaction was finalized.
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) finalizeTx(
+	ctx context.Context, atx AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+	tx, err := tr.txStore.GetTxByID(ctx, atx.id)
+	if err != nil {
+		tr.lggr.Errorw(errors.Wrap(err, "failed to get tx by ID").Error())
+		return false
+	}
+
+	finalized := false
+	switch tx.State {
+	case TxConfirmed, TxConfirmedMissingReceipt, TxFatalError:
+		finalized = true
+	case TxInProgress, TxUnstarted, TxUnconfirmed:
+		if atx.isValid() {
+			break
+		}
+		if err := tr.finalizeFatal(ctx, tx); err != nil {
+			tr.lggr.Errorw(err.Error())
+			break
+		}
+		finalized = true
+	default:
+		tr.lggr.Panicw(fmt.Sprintf("unhandled transaction state: %v", tx.State))
+	}
+
+	return finalized
+}
+
+// finalizeFatal sets a transaction's state to fatal_error
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) finalizeFatal(
+	ctx context.Context,
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
+	tx.Error.SetValid(fmt.Sprintf(
+		"abandoned transaction exceeded time to live of %d hours", int(tr.ttl.Hours())))
+
+	tx.State = TxInProgress
+	err := tr.txStore.UpdateTxFatalError(ctx, tx)
```

Comment:
I think Fatal is a slightly different status, than what we want.
We want to mark it as Abandoned, as the TTL expired, and we gave up on the Tx.
It is possible the Tx still gets successfully included in the chain later on, but our TXM doesn't want to track it any longer.

This might sound more work, but could you please create a new state as TxAbandoned in this file:https://github.com/smartcontractkit/chainlink/blob/172d4a40d0b4582d95bbdaf8f0c157100b38b23f/common/txmgr/models.go

And then, create a new method in TxStore called UpdateTxAsAbandoned(). It should be very similar to existing txStore.UpdateTxFatalError method.
You just need to remove the check that existing state for Tx should be InProgress.

Reply:
After looking into adding `TxAbandoned` state to the database enum, it seems that there's no safe way to do this. Adding to the enum will require a duct tape style solution, and not be a good idea for migrations as there's no nice way to migrate back.